### ISSUE TITLE
[BUG] Parsing FASTA files - IndexError: list index out of range

### ISSUE THREAD

**Issue by tivdnbos** [wfondrie/mokapot#13]
Hi, 

I started testing the mokapot software after reading the manuscript on JPR, but run into a problem by the following code in the command line:
`mokapot --proteins my_concatenated_target_decoy.fasta my.pin`
The other options are default, including the decoy_ prefix. I tried this already in Percolator where it worked, so I assume it must be something in mokapot. 

I received the following error:
```
...
[INFO] Found 51556 PSMs.
[INFO]   - 27743 target PSMs and 23813 decoy PSMs detected.
[INFO] Protein-level confidence estimates enabled.
[INFO] Parsing FASTA files and digesting proteins...
Traceback (most recent call last):
  File "/home/compomics/anaconda3/bin/mokapot", line 10, in <module>
    sys.exit(main())
  File "/home/compomics/anaconda3/lib/python3.6/site-packages/mokapot/mokapot.py", line 74, in main
    decoy_prefix=config.decoy_prefix,
  File "/home/compomics/anaconda3/lib/python3.6/site-packages/mokapot/proteins.py", line 97, in __init__
    decoy_prefix=decoy_prefix,
  File "/home/compomics/anaconda3/lib/python3.6/site-packages/mokapot/proteins.py", line 251, in read_fasta
    prot, seq = _parse_protein(entry)
  File "/home/compomics/anaconda3/lib/python3.6/site-packages/mokapot/proteins.py", line 431, in _parse_protein
    seq = entry[1].replace("\n", "")
IndexError: list index out of range
```

I tried to run it without the protein parameter, which worked. However, a few recommendations:
- the proteins that are listed in the .psms.txt file are separated with a tab, which makes it difficult to parse the results (e.g., awk to filter on q-value). Moreover, currently it's not consistent: if only protein could be identified, it's not marked between double quotes. This could be solved by moving the proteins column to the end (like Percolator .pout files), or by separating them by comma or semicolon (and dropping the double quotes)
- a minor thing, but it would be nice to see no spaces in the column headers (e.g MokapotScore, MokapotQvalue, MokapotPEP)

 As a side question, will the fast fido protein inference be implemented in the future?

Best wishes,
Tim

**Comment by tivdnbos** [comment]
Admittedly, I started of by explaining my bug, but gave you some featurerequests/improvements as well. Hope you don't mind it's in the same thread.

**Comment by wfondrie** [comment]
Hi Tim, 

First, thanks for giving mokapot a try! Also, no worries about having multiple things in one thread. Here are my responses:

> I received the following error

The error you're running into certainly appears to be a bug with mokapot's file parsing. Can you share an example target and decoy entry from your FASTA file (or the whole thing). I think I know what is going on, but I'd like to have an example to verify. We're planning on a new release later this week and I think we can resolve this issue quickly enough to deploy a fix as part of it.

> the proteins that are listed in the .psms.txt file are separated with a tab, which makes it difficult to parse the results (e.g., awk to filter on q-value). 

You bring up some good points. In general, mokapot makes no attempt to parse the protein column (it finds proteins based on the provided FASTA file for protein-level confidence estimation); it merely copies the protein column in the input file to the output. Thus, the protein column can take any form you'd like.

> Moreover, currently it's not consistent: if only protein could be identified, it's not marked between double quotes. 

This is an artifact of some proteins containing the delimited (`\t`) and using [Pandas](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_csv.html?highlight=to_csv#pandas.DataFrame.to_csv) to write our result files. The quotes around proteins that contain the delimiter allow for many tools to easily read the results file, so it it something that I'm not inclined to remove. For example, the result files can be read into Python simply with:

```Python
import pandas as pd
psms = pd.read_table("psms.txt")
```

In the case of awk, it looks like it is possible to respect these quotes: https://www.gnu.org/software/gawk/manual/html_node/Splitting-By-Content.html

> This could be solved by moving the proteins column to the end (like Percolator .pout files), or by separating them by comma or semicolon (and dropping the double quotes).

I'll look into moving the protein column to the end of the file. Unfortunately replacing the protein delimiter isn't really an option, because we don't impose restrictions as to how this column is formatted.

> a minor thing, but it would be nice to see no spaces in the column headers (e.g MokapotScore, MokapotQvalue, MokapotPEP)

This is something we thought about a good bit, before deciding to include column headers with spaces. We opted for names that included spaces for two reasons: (1) we think it's slightly clearer for users what the column means, and (2) most programs have mechanisms to handle spaces in column names these days. I know it can be a little annoying though. For example, in Python it mandates a specific syntax to retrieve the column for a Pandas DataFrame:

```Python
import pandas as pd
psms = pd.read_table("psms.txt")

# This syntax works for everything:
specid = psms["SpecId"]
qvals = psms["mokapot q-values"]

# This syntax does not work for names with spaces:
specid = psms.SpecId
qvals = psms.mokapot q-values
```

I hope these have answered your questions sufficiently. We'll the the FASTA bug fixed ASAP. Please let me know if you have any other questions or suggestions!

Best,
Will

**Comment by tivdnbos** [comment]
Hi Will, 

Thanks for your quick and elaborate reply!
I have pasted a part of the fasta file below (T and D of the same proteins).

> >sp|ID_ID_7_1_410| k127_7_1_410_+
> ADLIAEGRRRTVNIFEADKDKVIFENGRFAVTGTNYAIDFFELAKKTSQTDGALRVVRDNEMHTPVFPNGAAICEVEVDPDTGLVEITRYATVDDVGRCINPMIVHGQTHGGIAQGVGQAMWEQCYVEPWSGQPL
> >decoy_sp|ID_ID_7_1_410| k127_7_1_410_+
> LPQGSWPEVYCQEWMAQGVGQAIGGHTQGHVIMPNICRGVDDVTAYRTIEVLGTDPDVEVECIAAGNPFVPTHMENDRVVRLAGDTQSTKKALEFFDIAYNTGTVAFRGNEFIVKDKDAEFINVTRRRGEAILDA
> >sp|ID_8_1_428| k127_8_1_428_+
> YWLPAITSRELPNPDCPPIRVKLLSEELIAFRTTSGTVGLIQNACPHRGASLFFGRNEEDGLRCVYHGWKFDTAGNCVDMPSEPAESNFKNKVHARAYPTLEKAGLIWVYMGPPELVPELPNYGWSQVPDGHVHVSRRIQEC
> >decoy_sp|ID_8_1_428| k127_8_1_428_+
> CEQIRRSVHVHGDPVQSWGYNPLEPVLEPPGMYVWILGAKELTPYARAHVKNKFNSEAPESPMDVCNGATDFKWGHYVCRLGDEENRGFFLSAGRHPCANQILGVTGSTTRFAILEESLLKVRIPPCDPNPLERSTIAPLWY
> >sp|ID_9_1_381| k127_9_1_381_-
> EPPHVHVETAGKSAKCCLRPVALARSAGYDARELRVLRGLVEAHRDDLERAWHASTTPSPTVAPRPAGAQATAVWFAGDLLWVRLADGRELGAPLEWFPRLAGAAGAQRDRWELTDQGFGVHWPEL
> >decoy_sp|ID_9_1_381| k127_9_1_381_-
> LEPWHVGFGQDTLEWRDRQAGAAGALRPFWELPAGLERGDALRVWLLDGAFWVATAQAGAPRPAVTPSPTTSAHWARELDDRHAEVLGRLVRLERADYGASRALAVPRLCCKASKGATEVHVHPPE
> >sp|ID_10_1_380| k127_10_1_380_+
> TKDIPIIFLSAVNIDKRFITKGYNSGGADYITKPFDPDLLLLKVKTFCRLFEQNRDLKVVQNALREEIDSRKMAQQEVERMNSSLEEKVTQRTQELVQLNQDLETRNAELAQYASLASHDLQEPLR
> >decoy_sp|ID_10_1_380| k127_10_1_380_+
> RLPEQLDHSALSAYQALEANRTELDQNLQVLEQTRQTVKEELSSNMREVEQQAMKRSDIEERLANQVVKLDRNQEFLRCFTKVKLLLLDPDFPKTIYDAGGSNYGKTIFRKDINVASLFIIPIDKT

I haven't tried mokapot in Python, but will give it a try tomorrow and let you know if I encounter other bugs.

Best,
Tim

**Comment by wfondrie** [comment]
Thank you for sharing those. When I save these entries to a FASTA file, I don't have any issue.

Are you able to reproduce the error if you create a FASTA file with only these entries?

**Comment by wfondrie** [comment]
This fix is in the 0.6.0 release.
